### DISCUSSION TITLE
Caching of artifacts

### DISCUSSION THREAD

**Discussion by fonkadelic** [MobileNativeFoundation/discussions#40] [General]
Hi everyone,
would love to hear your strategies when it comes to caching artifacts to speed up builds.
From my experience [Rome](https://github.com/tmspzz/Rome) can give you quite a boost in build times in case [Carthage](https://github.com/Carthage/Carthage) is used as a dependency manager.
Since the Swift Package Manger is getting more and more adopted i'm wondering if there are approaches to use some sort of caching here as well.
A more radical thing to do is of course using a different build tool like [Bazel](https://bazel.build) that has caching features baked in.

**Comment by zachgrayio** [comment]
>A more radical thing to do is of course using a different build tool like Bazel that has caching features baked in.

Among many MNF members, I think you'll find adoption of these tools is extremely high, probably universal (not any _one_ tool per se, but tools in this class with the same feature set of hermeticity, reproducibility, shared artifact caching, remote build workers, and so on).

Of course, that doesn't mean it's worthwhile to deploy them for every project, but there definitely comes a point of scale at which teams are forced to face the problems these tools solve and it's becoming less and less advisable to begin this effort from scratch. 

Also it's worth noting that Gradle does have caching functionality out of the box these days, and that's becoming "par for the course" for any modern build tooling and ofc tools like distcc/icecream/sccache provide a lot of this as well and have been around for awhile.

I'll leave it to other folks to chime in here, but a shared artifact cache (which does include caching of 3rd-party deps in most cases depending on your vendoring strategy) is one of the universal features that almost any large team will either already have deployed, or is working towards, and remotely executed build actions also is seeing big growth. 

I wouldn't call the tooling radical at all 🙂

**Reply by dalemyers** [reply]
Yeah, I'm in agreement. All the larger teams I'm aware of already, as you said, use something like Bazel or are working on it. I'm on Outlook for iOS, and we are currently in the process of moving to Bazel. At a certain point running builds without caching just becomes untenable. For smaller projects, Rome works well, but we have found that Carthage in general isn't giving us the flexibility we need. A simpler alternative for smaller projects is to just check in your build Carthage dependencies (using git LFS), and then only pay the linking cost. However, that won't help you if your app is on the larger side. 

Agree again that Bazel isn't radical these days. The support for iOS apps is growing rapidly for it. It's only a matter of time before almost all large apps are using it (or similar).

**Reply by zachgrayio** [reply]
Started a thread on the topic here https://github.com/MobileNativeFoundation/discussions/discussions/41

**Reply by bharathns** [reply]
@dalemyers  how about Buck?

**Reply by dalemyers** [reply]
I would love to say I have a solid opinion on Buck, but the simple answer is that I don't. Several years ago (maybe 4 now?), one of our most respected devs looked at it and decided it wasn't for us. He really knew his stuff so I have no reason to look at it again, even though 4 years have passed. I'm sure that on average, either is fine, but he made the call that Bazel is better, so we've run with it for now.

**Comment by keith** [comment]
Outside of the bazel path, which is also what we do, before we were on it we also used pre-compiled binary dependencies for developer builds. Our path was pretty much to build them as static frameworks, vendor them via s3 (as mentioned above I think git-lfs would have had similar results we just haven't ever used it), and include them with XcodeGen. This was a huge productivity win to avoid potential re-compiles of things developers never intentionally changed.

I hesitate to recommend what we were doing to build them, but the most similar option out there to it is if you actually used a tool like bazel _just for that_ and created a [`ios_static_framework`](https://github.com/bazelbuild/rules_apple/blob/master/doc/rules-ios.md#ios_static_framework) rule for each framework. This might be too cumbersome if you have heavily nested dependency trees, but we didn't so it work well for us for years.

**Reply by bharathns** [reply]
But does it involve multiple checkins if we do some changes in those frameworks?

**Reply by keith** [reply]
I'm not sure I follow what you're asking there but if something changed in our s3 case we would swap symlinks between the new and the old version when developers regenerated their project (which was always a pre-req for our setup when people checked out a new sha, even when we were using cocoapods so we could run `pod install`). With git-lfs I guess this would "just work"

**Reply by bharathns** [reply]
@keith  what I wanted to know is..
When there is a change in one of those module which you build and store in s3, what is the process you follow to update the cache? Mostly wanted to know what was done here -> "I hesitate to recommend what we were doing to build them"
1. Would it be separate build for those frameworks which would update the s3
2. Or would it be single build where there is a cache miss which would trigger the full build and update the s3.

**Reply by keith** [reply]
Ah got it, when someone updates a library they would manually bump whatever versions, build it locally, upload it to s3, and merge a change that points to the new s3 version. It can be a bit involved but given that developers build way more often than we update dependencies I think the tradeoff is worth it. Of course bazel has the ideal solution here where you have a build cache for all artifacts that will build or pull from the cache depending on whether or not it has been built before

**Reply by bharathns** [reply]
NIce! Thanks that makes it clear.

**Comment by vito1188** [comment]
> From my experience [Rome](https://github.com/tmspzz/Rome) can give you quite a boost in build times in case [Carthage](https://github.com/Carthage/Carthage) is used as a dependency manager.

My team is using Rome for Carthage dependencies and [cocoapods-binary](https://github.com/leavez/cocoapods-binary) for Cocoapods dependencies.

**Comment by shepting** [comment]
Yeah, caching is super valuable. A clean Xcode build for us is 45 long, a clean Buck build for us is 19 minutes, and a Buck build with HTTP cache averages 4.5 minutes (no tests, just building in every case).

**Comment by tinder-maxwellelliott** [comment]
So I experimented around with this at Tinder, using https://github.com/grab/cocoapods-binary-cache and quickly realized that this was untenable for several reasons:

1. Developers expect to build for anything at any time, meaning a developer can go from Debug to Release and then to Device all in a session. Managing all of those prebuilt artifacts is a significant challenge. Also developers **expect projects to open as quickly as they do currently** so downloading all of these artifacts is not acceptable.
2. We did not see a big performance win from caching our 3rd party dependencies, many of them were vendored frameworks to begin with or were very small, recompilation was not a significant issue. It was clear that the internal modules were the issue.
3. Even in this world of pre-compilation we still had to test everything in CI, it was performance capped from the start. 

Given those considerations we diverged from this route (_thank goodness_) and adopted Bazel alongside our tool https://github.com/Tinder/bazel-diff. It is truly impressive to get to the point where you can skip CI or only run a single suite of tests for a change, alongside the huge stability gains. That is where you want to be, not farming over s3 buckets of frameworks constantly for marginal gain.

**Comment by polac24** [comment]
[iOS] At Spotify we are using https://github.com/spotify/XCRemoteCache to cache first-party targets - we have more than 400 of them. The tool seamlessly integrates into existing .xcodeproj as an extra build step and skips local compilation if the remote artifact is available on a remote cache server.
The main benefit is its "add-on" nature - it requires minimal effort to integrate and maintain plus the Xcode IDE experience is identical with the standard flow. Technically, that tool can be used for third-party dependencies for we enabled that only for local builds and internal targets.
